enhancements/cluster-logging/forward_to_cloudwatch.md

```diff
@@ -0,0 +1,198 @@
+---
+title: forward_to_cloudwatch
+authors:
+  - "@alanconway"
+reviewers:
+  - "@jcantrill"
+  - "@jeremyeder"
+approvers:
+creation-date: 2020-12-17
+last-updated: 2020-12-17
+status: implementable
+see-also:
+superseded-by:
+---
+
+# Forward to Cloudwatch
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+[Amazon Cloudwatch][aws-cw] is a hosted monitoring and log storage service.
+This proposal extends the `ClusterLogForwarder` API with an output type for Cloudwatch.
+
+## Motivation
+
+Amazon Cloudwatch is a popular log store.
+We have requests from external and  Red Hat-internal customers to support it.
+
+### Goals
+
+Enable log forwarding to Cloudwatch.
+
+### Non-Goals
+
+Enable Cloudwatch metric collection.
+
+## Proposal
+
+New API fields in the `output.cloudwatch` section:
+
+- `region`: (string) AWS region name.
+- `retentionDays`: (number) Number of days to keep logs. Has a default.
```

Comment:
Is the intention that the forwarder owns the CloudWatch instance and sets it up?  If so, then this parameter makes sense.

But if the forwarder is being used to send logs to an instance that is externally managed, what is the log forwarder going to do with this parameter?  Reach out and change the CloudWatch instance to have a retention time that matches this parameter?

Reply:
If the forwarder doesn't have permission to manage this and it's externally managed CW would this just not be set and that bit of config skipped?  If not set, do nothing.

Reply:
This API was worked back from the options available by both the plugin and the Cloudwatch API.  The plugin allows setting this parameter which presumably means it will config auto-created loggroups and logstreams with this value.  The intention is up for debate but the very early prototype assumes the forwarder will create the logroup and logstream under the guidance the only decision ROSA users will make is to opt-into Cloudwatch (e.g. checkbox, true/false)

I don't know the behavior of the CW api so unable to speak to the ramifications of setting the value for externally managed CW.  Seems this would be tangential however as credentials provided, at least in the ROSA case, presumably will be able to access and control whatever loggroup is created by the forwarder

Reply:
The fluentd plugin has this option, and from my reading of the cloudwatch it appears that cloudwatch allows the collector to set retention times, even on individual log entries, which surprised me. Possibly I have misinterpreted? It seems that the collector drives a lot of the cloudwatch configuration, creating groups and so on.

Reply:
I've clarified the subset of stuff that's for first release.
As I read the docs, it is the collector that sets up cloudwatch: creates and deletes groups and streams, sets retentions (even on a per-entry basis??!) etc. Shout if I misread.
Does anyone have a working cloudwatch config we can look at?